### ISSUE TITLE
Auto-discovery of node-local storage devices

### ISSUE THREAD

**Issue by SteVwonder** [flux-framework/flux-sched#651]
For systems like Corona, Lassen, and Catalyst, we would benefit from the automatic discovery of node-local storage devices.  It appears that hwloc already supports this, but when I run `lstopo` on Corona compute nodes, I don't see any storage devices.  I suspect that the system hwloc is either too old or not configured with the proper dependencies to find them.  If it is a version issue, then this is blocked by https://github.com/flux-framework/flux-core/issues/1377.  If it's just a configuration issue, then that is a simpler fix.

I'll report back with what I find and go from there.

**Comment by dongahn** [comment]
Does `--whole-io` option help at all?

**Comment by SteVwonder** [comment]
> Does --whole-io option help at all?

Unfortunately, it did not.  The NVMe drive is successfully detected (without `--whole-io` or any other special flags) after installing v2.1 via Spack.  Not sure if it is just a version thing or a configuration thing.  Digging deeper.

```
15s herbein1@corona147 ~/workspace/storage-auto-discovery
❯ lstopo --version                                                                   17:26:00 ()
lstopo 2.1.0
                                                                                                 
herbein1@corona147 ~/workspace/storage-auto-discovery
❯ lstopo | grep -i nvme                                                              17:26:05 ()
          PCI 01:00.0 (NVMExp)
            Block(Disk) "nvme0n1"
                                                                                                 
herbein1@corona147 ~/workspace/storage-auto-discovery
❯ lstopo | grep -i disk                                                              17:26:13 ()
            Block(Disk) "nvme0n1"
  Block(Disk) "sda"
                                                                                                 
herbein1@corona147 ~/workspace/storage-auto-discovery
❯ spack unload hwloc                                                                 17:26:15 ()
                                                                                                 
18s herbein1@corona147 ~/workspace/storage-auto-discovery
❯ lstopo --version                                                                   17:26:57 ()
lstopo 1.11.8
                                                                                                 
herbein1@corona147 ~/workspace/storage-auto-discovery
❯ lstopo | grep -i nvme                                                              17:27:08 ()
                                                                                                 
(1) herbein1@corona147 ~/workspace/storage-auto-discovery
❯ lstopo | grep -i disk       
```

Interesting wrinkle: hwloc is also picking up the disk at `/dev/sda`, which doesn't seem to be mounted on this machine.  Here is the extra metadata available at the XML output level:

```
❯ lstopo --of xml | grep -A8 -i disk                                                                                                                                                                                                                                                      17:31:46 ()
              <object type="OSDev" gp_index="384" name="nvme0n1" subtype="Disk" osdev_type="0">
                <info name="Size" value="1562813784"/>
                <info name="SectorSize" value="4096"/>
                <info name="LinuxDeviceID" value="259:0"/>
                <info name="Model" value="HUSMR7616BDP301"/>
                <info name="SerialNumber" value="SDM000023759"/>
              </object>
            </object>
          </object>
--
    <object type="OSDev" gp_index="383" name="sda" subtype="Disk" osdev_type="0">
      <info name="Size" value="26214400"/>
      <info name="SectorSize" value="512"/>
      <info name="LinuxDeviceID" value="8:0"/>
      <info name="Vendor" value="LIO-ORG"/>
      <info name="Model" value="FILEIO"/>
      <info name="Revision" value="4.0"/>
      <info name="SerialNumber" value="60014054e2824b7073e3300794400382"/>
    </object>

```

Unfortunately, I don't see anything obvious to differentiate these devices as "useful" vs "not useful" for the user.  Maybe more information is available at the API level.  It is too early to say for sure, but we may need to add some extra logic/configuration options such as an allow/block list to filter out the disks we do not care about and keep the ones we do (configurable per machine).  If there is a way to get the mountpath from hwloc, maybe we do some extra checks, like if the directory is writeable.

**Comment by SteVwonder** [comment]
I tried a few combinations of versions and runtime flags.  Ending up opening an issue upstream asking for clarification on the minimum version that provides the functionality that we need: https://github.com/open-mpi/hwloc/issues/399

AFAICT, the 2.X series is the first to support both detecting NVMe devices *AND* reporting their size.  The RHEL provided 1.11.8 cannot detect them, and the 1.11.11 version detects NVMe devices but does not report their size.  Well see what Brice says.

**Comment by SteVwonder** [comment]
Per the comment from Brice in the linked issue, it looks like:

1. RHEL is backporting changes into their kernel which means that NVMe devices will not show at all in hwloc < 1.11.10
2. The size of detected NVMe devices is only reported in the 2.X branch of hwloc

So I believe we need to handle https://github.com/flux-framework/flux-core/issues/1377 before we can reliably auto-detect node-local storage.

**Comment by dongahn** [comment]
Given that hwloc v2+ support has been merged into both flux-core and fluxion, this should be handled correctly. Closing. If it is not the case, we can aways reopen.

**Comment by SteVwonder** [comment]
One small PR remains before I think we can close this.  There needs to be a new case added to the big switch statement in the hwloc reader, and then a test case added using an hwloc file from corona or lassen.

**Comment by dongahn** [comment]
Ah. Makes sense. Thanks.

**Comment by dongahn** [comment]
PR #714 has been merged!
